framework/src/modules/dpos/transaction_assets/unlock_transaction_asset.ts

```diff
@@ -105,25 +112,36 @@ export class UnlockTransactionAsset extends BaseAsset<UnlockTransactionAssetCont
 				throw new Error('Corresponding unlocking object not found.');
 			}
 
-			const waitingPeriod = getWaitingPeriod(
-				sender.address,
-				delegate.address,
-				store.chain.lastBlockHeaders[0].height,
-				unlock,
-			);
+			const currentHeight = store.chain.lastBlockHeaders[0].height + 1;
+			const minWaitingHeight = getMinWaitingHeight(sender.address, delegate.address, unlock);
 
-			if (waitingPeriod > 0) {
+			if (minWaitingHeight > currentHeight) {
 				throw new Error('Unlocking is not permitted as it is still within the waiting period.');
 			}
 
-			const punishmentPeriod = getPunishmentPeriod(
-				sender,
-				delegate,
-				store.chain.lastBlockHeaders[0].height,
-			);
+			const minUnlockDelay = getMinPunishedHeight(sender, delegate);
+			const lastPomHeight =
+				delegate.dpos.delegate.pomHeights.length === 0
+					? undefined
+					: Math.max(...delegate.dpos.delegate.pomHeights);
 
-			if (punishmentPeriod > 0) {
-				throw new Error('Unlocking is not permitted as the delegate is currently being punished.');
+			if (currentHeight >= this._unlockFixHeight) {
+				if (
+					minUnlockDelay > currentHeight &&
+					lastPomHeight !== undefined &&
+					lastPomHeight < minWaitingHeight
+				) {
+					throw new Error(
+						'Unlocking is not permitted as the delegate is currently being punished.',
+					);
+				}
+				// unlocking condition before hard fork, to be deprecated
+			} else if (currentHeight < this._unlockFixHeight) {
+				if (minUnlockDelay > currentHeight) {
+					throw new Error(
+						'Unlocking is not permitted as the delegate is currently being punished.',
+					);
+				}
```

Comment:
```suggestion
			if (currentHeight >= this._unlockFixHeight &&
					minUnlockDelay > currentHeight &&
					lastPomHeight !== undefined &&
					lastPomHeight < minWaitingHeight
			) {
				throw new Error(
					'Unlocking is not permitted as the delegate is currently being punished.',
				);
			}
			// unlocking condition before hard fork, to be deprecated
			if (currentHeight < this._unlockFixHeight && minUnlockDelay > currentHeight) {
				throw new Error(
					'Unlocking is not permitted as the delegate is currently being punished.',
				);
			}
```

I think nested if can be combined?